File: python/lib/morsel/sensors/range_camera.py
from panda3d.pandac.Modules import Point3
from panda3d.pandac.Modules import PNMImage
from panda3d.pandac.Modules import Texture
from panda3d.pandac.Modules import Vec4
from panda3d.pandac.Modules import VBase3

from panda3d.pandac.Modules import BitMask32
from panda3d.pandac.Modules import PandaNode
from panda3d.pandac.Modules import NodePath

from math import *

class RangeCamera( NodePath ):
  def __init__( 
    self, 
    name, 
    hFov = 90, 
    vFov = 90, 
    hRayCount = 360, 
    vRayCount = 360, 
    rayRange = [1, 10]
  ) :
    NodePath.__init__( self, name )
    self.width  = 128
    self.height = 128
    self.hFov = hFov
    self.vFov = vFov
    self.hRayCount = hRayCount
    self.vRayCount = vRayCount
    self.rayRange = rayRange
    self.precomp  = []
    self.rays     = [[]] * (self.hRayCount * self.vRayCount)
    self.rayCount = self.hRayCount * self.vRayCount
    self.setupCamera()
    self.setupRays()
    
#-------------------------------------------------------------------------------

  def update( self, time ):
    #print self.depthmap.get_size_x()
    self.depthmap.store( self.texels )
    self.updateRays()
    return True

#-------------------------------------------------------------------------------

  def setupCamera( self ):
    hStartAngle      = self.hFov / 2.0
    vStartAngle      = self.vFov / 2.0
    self.texels      = PNMImage()
    self.depthmap    = Texture() 
    
    self.depthmap.setFormat( Texture.FDepthComponent )
    self.depthmap.setComponentType( Texture.TUnsignedShort )
    
    self.depthbuffer = base.win.makeTextureBuffer( 'depthmap', self.width, self.height, self.depthmap, True )
    self.camera      = base.makeCamera( self.depthbuffer )
    
    self.camera.reparentTo( self )
    
    self.camera.node().setScene( render )
    self.camera.node().getLens().setNearFar( *self.rayRange )
    self.camera.setHpr( -90, 0, 180 )
    self.camera.node().getLens().setFov( self.hFov )
    self.camera.node().getLens().setAspectRatio( 1.0 * self.hFov / self.vFov )
    self.camera.node().setCameraMask( BitMask32( '001' ) )
    self.camera.setLightOff( 1 )
    self.camera.setMaterialOff( 1 )
    self.camera.setColorOff( 1 )

#-------------------------------------------------------------------------------

  def setupRays( self ):
    hFov        = self.hFov * pi / 180
    vFov        = self.vFov * pi / 180
    if self.hRayCount > 1:
      deltaHAngle = -hFov / (self.hRayCount - 1 )
      hAngle      = hFov / 2
    else:
      deltaHAngle = -hFov
      hAngle      = 0
    if self.vRayCount > 1:
      deltaVAngle = -vFov / (self.vRayCount - 1 )
    else:
      deltaVAngle = -vFov
    hFocal = ( self.width / 2 - 0.5) / tan( hFov / 2 )
    vFocal = ( self.height / 2 - 0.5) / tan( vFov / 2 )
    while hAngle + hFov / 2 > -2e-10:
      hDist = hFocal * tan( hAngle )
      col   = int( self.width / 2 + hDist )
      if self.vRayCount > 1:
        vAngle      = vFov / 2
      else:
        vAngle      = 0
      while vAngle + vFov / 2 > -2e-10:
        vDist = vFocal * tan( vAngle )
        row   = int( self.height / 2 + vDist )
        self.precomp.append( [
          hAngle, vAngle, 
          col, row, 
          tan( hAngle ), tan( vAngle ), 
          hAngle,
          vAngle,
        ] )
        vAngle += deltaVAngle
      hAngle += deltaHAngle
      
#-------------------------------------------------------------------------------
    
  def updateRays( self ):
    i = 0
    for rayInfo in self.precomp:
      col    = rayInfo[2]
      row    = rayInfo[3]
      depth   = self.texels.getGray( col, row )
      x = self.rayRange[1] * self.rayRange[0] / ( self.rayRange[1] - depth * ( self.rayRange[1] - self.rayRange[0] ) )
      if x < 0:
        x = 0
      y = x * rayInfo[4]
      z = x * rayInfo[5]
      r = sqrt( x*x + y*y + z*z )
      self.rays[i] = [col, row, x, y, z]
      i += 1

#-------------------------------------------------------------------------------

  def inRange( self, node ):
    ''' Returns true if node is in the sensor's range.
    TODO: Take into account FOV.
    '''
    p = self.getRelativePoint( node, VBase3( 0, 0, 0 ) )
    distance = sqrt( p.dot( p ) )
    if distance >= self.rayRange[0] and distance <= self.rayRange[1]:
      if p[0] > 0:
        return True
    return False

File: src/lib/morsel/sensors/range_camera.cpp
/***************************************************************************
 *   Copyright (C) 2011 by Ralf Kaestner                                   *
 *   ralf.kaestner@gmail.com                                               *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include "range_camera.h"

#include "morsel/utils/color.h"

#include <cmath>
#include <string>
#include <limits>

using namespace std;

/*****************************************************************************/
/* Constructors and Destructor                                               */
/*****************************************************************************/

RangeCamera::RangeCamera(const string& name, const ShaderProgram& program,
    const LVecBase2f& angles, const LVecBase2f& fov, const LVecBase2f&
    numRays, const LVecBase2f& rangeLimits, const LVecBase2f& resolution,
    bool acquireColor, string acquireLabel) :
  NodePath(name),
  program(program),
  angles(angles),
  fov(fov),
  numRays(numRays),
  rangeLimits(rangeLimits),
  resolution(resolution),
  acquireColor(acquireColor),
  acquireLabel(acquireLabel),
  rays(numRays[0]*numRays[1]),
  timestamp(0.0),
  lastDepthTimestamp(0.0),
  lastColorTimestamp(0.0),
  lastLabelTimestamp(0.0) {
}

RangeCamera::~RangeCamera() {
}

/*****************************************************************************/
/* Accessors                                                                 */
/*****************************************************************************/

const Lens& RangeCamera::getLens() const {
  return *cameraNode->get_lens();
}

const Texture& RangeCamera::getDepthMap() const {
  return depthMap;
}

const Texture& RangeCamera::getColorMap() const {
  return colorMap;
}

const Texture& RangeCamera::getLabelMap() const {
  return labelMap;
}

size_t RangeCamera::getNumRays() const {
  return numRays[0]*numRays[1];
}

size_t RangeCamera::getNumHorizontalRays() const {
  return numRays[0];
}

size_t RangeCamera::getNumVerticalRays() const {
  return numRays[1];
}

const RangeCamera::Ray& RangeCamera::getRay(int index) const {
  return rays[index];
}

double RangeCamera::getTimestamp() const {
  return timestamp;
}

double RangeCamera::getDepth(int index) const {
  const Ray& ray = rays[index];
  
  if (timestamp > lastDepthTimestamp) {
    depthMap.store(depthTexels);
    lastDepthTimestamp = timestamp;
  }
  
  return rangeLimits[1]*rangeLimits[0]/(rangeLimits[1]-depthTexels.get_gray(
    ray.column, ray.row)*(rangeLimits[1]-rangeLimits[0]));
}

LPoint3f RangeCamera::getPoint(int index) const {
  const Ray& ray = rays[index];
  LPoint3f point;

  point[1] = getDepth(index);
  point[0] = point[1]*ray.hTan;
  point[2] = point[1]*ray.vTan;

  return point;
}

Colorf RangeCamera::getColor(int index) const {
  const Ray& ray = rays[index];
  Colorf color;
  
  if (timestamp > lastColorTimestamp) {
    colorMap.store(colorTexels);
    lastColorTimestamp = timestamp;
  }

  color[0] = colorTexels.get_red(ray.column, ray.row);
  color[1] = colorTexels.get_green(ray.column, ray.row);
  color[2] = colorTexels.get_blue(ray.column, ray.row);
  color[3] = 1.0;

  return color;
}

size_t RangeCamera::getLabel(int index) const {
  const Ray& ray = rays[index];
  
  if (timestamp > lastLabelTimestamp) {
    labelMap.store(labelTexels);
    lastLabelTimestamp = timestamp;
  }

  return Color::rgbToInt(labelTexels.get_xel_a(ray.column, ray.row));
}

Colorf RangeCamera::getLabelColor(int index) const {
  const Ray& ray = rays[index];
  Colorf color;

  if (timestamp > lastLabelTimestamp) {
    labelMap.store(labelTexels);
    lastLabelTimestamp = timestamp;
  }

  color[0] = labelTexels.get_red(ray.column, ray.row);
  color[1] = labelTexels.get_green(ray.column, ray.row);
  color[2] = labelTexels.get_blue(ray.column, ray.row);
  color[3] = 1.0;

  return color;
}

void RangeCamera::setActive(bool active) {
  cameraNode->set_active(active);
}

/*****************************************************************************/
/* Methods                                                                   */
/*****************************************************************************/

bool RangeCamera::update(double time) {
  timestamp = time;
  return true;
}

void RangeCamera::showFrustum() {
  cameraNode->show_frustum();
}

void RangeCamera::hideFrustum() {
  cameraNode->hide_frustum();
}
                                          
void RangeCamera::setupCamera(PointerTo<Lens> lens) {
  depthMap.set_format(Texture::F_depth_component);
  depthMap.set_component_type(Texture::T_unsigned_short);
  depthMap.set_minfilter(Texture::FT_nearest);
  depthMap.set_magfilter(Texture::FT_nearest);

  buffer = Morsel::getWindow(0)->make_texture_buffer("DepthMap",
    resolution[0], resolution[1], &depthMap, true);

  cameraNode = new Camera("Camera");
  cameraNode->set_camera_mask(BitMask32(1));
  cameraNode->set_scene(Morsel::getGSG()->get_scene()->get_scene_root());
  cameraNode->set_lens(lens);

  camera = attach_new_node(cameraNode);
  camera.set_light_off(numeric_limits<int>::max());
  camera.set_material_off(numeric_limits<int>::max());
  camera.set_color_off(numeric_limits<int>::max());
  camera.set_transparency(TransparencyAttrib::M_none);

  PointerTo<DisplayRegion> drd = buffer->make_display_region();
  drd->set_sort(0);
  drd->set_camera(camera);
  
  if (acquireColor) {
    colorMap.set_minfilter(Texture::FT_nearest);
    colorMap.set_magfilter(Texture::FT_nearest);
    buffer->add_render_texture(&colorMap, GraphicsOutput::RTM_copy_ram,
      DrawableRegion::RTP_color);
  }
  else if (!acquireLabel.empty()) {
    labelMap.set_minfilter(Texture::FT_nearest);
    labelMap.set_magfilter(Texture::FT_nearest);
    buffer->add_render_texture(&labelMap, GraphicsOutput::RTM_copy_ram,
      DrawableRegion::RTP_color);

    program.define("LABEL", acquireLabel);
    shader = program.make();

    NodePath shaderAttrib("ShaderLabel");
    shaderAttrib.set_shader(shader);
    shaderAttrib.set_shader_input(acquireLabel, Color::intToRgb(0));
    cameraNode->set_initial_state(shaderAttrib.get_state());
    buffer->set_clear_color(Color::intToRgb(0));
  }
}
